src/wp-includes/html/class-wp-html-attribute-token.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * HTML Tag Processor: Attribute token structure class.
+ *
+ * @package WordPress
+ * @subpackage HTML
+ * @since 6.2.0
+ */
+
+/**
+ * Data structure for the attribute token that allows to drastically improve performance.
+ *
+ * This class is for internal usage of the WP_HTML_Tag_Processor class.
+ *
+ * @access private
+ * @since 6.2.0
+ *
+ * @see WP_HTML_Tag_Processor
+ */
+final class WP_HTML_Attribute_Token {
+	/**
+	 * Attribute name.
+	 *
+	 * @since 6.2.0
+	 * @var string
+	 */
+	public $name;
+
+	/**
+	 * Attribute value.
+	 *
+	 * @since 6.2.0
+	 * @var int
+	 */
+	public $value_starts_at;
+
+	/**
+	 * How many bytes the value occupies in the input HTML.
+	 *
+	 * @since 6.2.0
+	 * @var int
+	 */
+	public $value_length;
+
+	/**
+	 * The string offset where the attribute name starts.
+	 *
+	 * @since 6.2.0
+	 * @var int
+	 */
+	public $start;
+
+	/**
+	 * The string offset after the attribute value or its name.
+	 *
+	 * @since 6.2.0
+	 * @var int
+	 */
+	public $end;
+
+	/**
+	 * Whether the attribute is a boolean attribute with value `true`.
+	 *
+	 * @since 6.2.0
+	 * @var bool
+	 */
+	public $is_true;
```

Comment:
I agree with @anton-vlasenko. Why expose all of the properties as `public`? Does the API need all of them to be `public`?

Instead, only expose what is essential now. Make the reset `protected`.

Why?

All `public` properties and methods must remain publicly exposed for BC. But a `protected` method or property can later be changed to `public` if there's a need.

So for BC, I'd suggest:
* Evaluating each property and method to determine if it's a must-have for each to be publicly exposed.
* Change any to `protected` that are not a must-have.

In doing so, this limits the BC exposure for future maintenance.

Reply:
Great points! `const` should also be considered in this, as we're not currently at a PHP 7.1 minimum and can't indicate visibility. _**If**_ these properties/constants don't need to be publicly accessible, we might consider `private` or `protected` properties instead - whichever is appropriate.

Reply:
Looking more deeply at these modeling classes (ie `WP_HTML_Span`, `WP_HTML_Text_Replacement`, and `WP_HTML_Attribute_Token`, some more thoughts:

* In order to access each property within the processor, getters or specific methods would be needed.
* Too risky to change the architecture at this point.
* The objects are `private` properties within the processor. That will prevent unnecessary or unexpected state changes.

I'm thinking the design is okay as is.

Reply:
> Too risky to change the architecture at this point.

I agree and I absolutely understand it, @hellofromtonya.
It would be nice to have them as private (it's a final class so it cannot be extended), but I don't expect @dmsnell to refactor these properties.
I'm just curious why the decision was made to make these properties public.
cc @adamziel

Reply:
I wanted to drop this [comment](https://github.com/WordPress/wordpress-develop/pull/3920#discussion_r1091737197) again in case it was missed. I don't think we should make these classes `final` for extensibility. That of course doesn't mean that we shouldn't make the props private or protected.

Reply:
repeating here since I answered above before seeing this more lengthy discussion.

these are filling in for associate arrays. we wouldn't be having this discussion if instead we saw `$token['start']` and `$token['end']`. the arbitrary restriction to `private` or `protected`, and the addition of getters and setters, would only add to the needless bloat of the class without serving a purpose.

these classes are records; they hold and return data. if PHP had a record type or a tuple type or an efficient array/vector type these would be those, but as it stands the switch from associate array to class brought a major uplift in performance, including a substantial drop in the allocated memory use.

I'm _strongly_ opposed to changing these to protected or private due to the implications for direct access that they were created for, as well as the push towards bloat and stat e mentality.

Reply:
I don't see the need at this point in changing the design. See my reasoning here https://github.com/WordPress/wordpress-develop/pull/3920/#discussion_r1092085391.

Overall, there's no risk of this models being publicly exposed. Why? They are not publicly exposed within the tag processor.

There would be too much risk to modify the architecture at this late stage.

So with that, I'll close this thread as resolved. But please feel free to reopen it if more discussion is needed.

Reply:
> the switch from associate array to class brought a major uplift in performance, including a substantial drop in the allocated memory use.

That answers my question. Thanks!

> I'm strongly opposed to changing these to protected or private 

And I'm not expecting you to change it.
I was just curious why these properties were declared as public as it breaks encapsulation.

> and the addition of getters and setters, would only add to the needless bloat of the class without serving a purpose.

The purpose is to prevent accidentally changing values assigned to the properties.

Reply:
> I was just curious why these properties were declared as public as it breaks encapsulation.

while I understand where you're coming from with respect to the idea of encapsulation, that principle in itself can't be a blunt tool to wipe out any and all public properties. if it were so, nobody would have created public properties.

if you'd like to hold a more in-depth discussion of object-oriented mentality feel free to poke me in slack as I'd be happy to share and listen. of note, the move from a property to getters and setters for that property is itself extremely counter to some of the central ideas in object orientation.

yes, these aren't being encapsulated, but they aren't supposed to be either. so when _wanting_ to encapsulate information it might be inappropriate to declare public visibility, but here I don't find any need to do so, and the class is fairly limited and self-evident. on the contrary, if these classes contained some abstract value which depended on any kind of complicated computation from its inputs then I would see more value in attempting to hide that, but still, would try to do so with a different approach than adding getters and setters as those carry no semantic value.

on a fun note we considered naming these even more abstractly given that they are literally nothing more than a 2-tuple, a 3-tuple, and a 5-tuple, but that didn't seem to jive as much with how most systems are more concretely built inside WordPress.

the block parser also utilizes helper classes (again for performance reasons) and actually does hide those classes within the bigger parser file. in the case of the tag processor, given its size, we decided to separate them out, knowing that in doing so it would invite questions about turning them more into what many people think of when they see "class"

Reply:
Thanks for the detailed reply.

> if you'd like to hold a more in-depth discussion of object-oriented mentality feel free to poke me in slack a

Thanks for offering it.
Although I disagree that getters and setters carry no semantic value, I don't see the need to discuss anything extra in Slack. I'm fine with your explanation.